cpp/include/cudf/detail/aggregation/aggregation.hpp

```diff
@@ -27,6 +27,63 @@
 
 namespace cudf {
 namespace detail {
+
+class aggregation_finalizer;
+
+class compound_aggregation {
+ public:
+  virtual std::vector<aggregation> get_simple_aggregations(data_type col_type) const { return {}; }
+  virtual void finalize(aggregation_finalizer& finalizer) = 0;
+};
+
+// Forward declare compound aggregations.
+class mean_aggregation;
+class std_var_aggregation;
+class min_aggregation;
+class max_aggregation;
+
+// Visitor pattern
+class aggregation_finalizer {  // Declares the interface for the finalizer
+ public:
+  // Declare overloads for each kind of a agg to dispatch
+  virtual void Dispatch(min_aggregation& agg)     = 0;
+  virtual void Dispatch(max_aggregation& agg)     = 0;
+  virtual void Dispatch(mean_aggregation& agg)    = 0;
+  virtual void Dispatch(std_var_aggregation& agg) = 0;
+};
+
+/**
+ * @brief Derived class for specifying a min aggregation
+ */
+struct min_aggregation final : compound_aggregation, aggregation {
+  min_aggregation(aggregation::Kind k) : compound_aggregation{}, aggregation{k} {}
```

Comment:
This ctor seems odd. This would allow constructing `min_aggregation(aggregtation::Kind::max)`.

The `kind` for `min_aggregation` should never be allowed to be anything other than `min`.

Reply:
That's the same case for `std_var_aggregation`, `nunique_aggregation`, `nth_element_aggregation`, `udf_aggregation` as well.
should they be designed to not allow any other `kind` value stored other than their own?

Reply:
> should they be designed to not allow any other kind value stored other than their own?

Yes. I'm not sure why this ctor needs to exist at all.

Reply:
fixed.